### PR TITLE
[SYCL-MLIR]: Memory access analysis (part1)

### PR DESCRIPTION
**Pull request by etiotto**
Given an affine loop, the access pattern of an array access can be modeled by using an access matrix and a vector of offsets. 
As an example, given the following loop
```
  for(i)
    for(j)
      A[c1*i+c2*j+c3][-c4*i-c5*j+c6]
```
The array access 'A' can be described using:
```
  |c1   c2|   |i|   |c3|
  |       | * | | + |  |
  |-c4 -c5|   |j|   |c6|
```
This PR introduces the `MemoryAccessMatrix` and `OffsetVector` classes which are designed to represents the above data structures.

### CODE REVIEW

File: llvm/include/llvm/ADT/APInt.h
Comment (whitneywhtsang) [inline llvm/include/llvm/ADT/APInt.h:1]: better not to change a LLVM file for formatting to avoid merge conflict.

File: polygeist/tools/cgeist/driver.cc
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/driver.cc:1]: please revert this change.

File: mlir-sycl/lib/Dialect/SYCL/Analysis/CMakeLists.txt [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
+  
   ADDITIONAL_HEADER_DIRS
   ${MLIR_SYCL_INCLUDE_DIR}/mlir/Analysis
 
   LINK_LIBS PUBLIC
-  MLIRSYCLDialect
+  MLIRAffineDialect
   MLIRAnalysis
+  MLIRArithDialect
+  MLIRMemRefDialect
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/SYCL/Analysis/CMakeLists.txt:12]: Are `MLIRArithDialect` and `MLIRMemRefDialect` used?

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+  os << "AccessMatrix:\n" << access.getAccessMatrix() << "\n";
+  os << "OffsetVector:\n" << access.getOffsetVector() << "\n";
+  os << "\n------------------\n";
+  return os;
+}
+
+} // namespace sycl
+} // namespace mlir
+
+#endif // MLIR_DIALECT_SYCL_ANALYSIS_MEMORYACCESSANALYSIS_H
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:421]: add new line.

**Review by whitneywhtsang — COMMENTED** [review]
A few obvious comments first.

File: mlir-sycl/unittests/Analysis/MemoryAccessAnalysis/MatrixTest.cpp [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+    }
+
+  EXPECT_THAT(matrix.isZero(solver), false);
+  EXPECT_THAT(matrix.isDiagonal(solver), false);
+  EXPECT_THAT(matrix.isIdentity(solver), false);
+  EXPECT_THAT(matrix.isLowerTriangular(solver), false);
+  EXPECT_THAT(matrix.isUpperTriangular(solver), true);
+}
+
+TEST(MatrixTest, PatterClassification) {
```
Comment (whitneywhtsang) [inline mlir-sycl/unittests/Analysis/MemoryAccessAnalysis/MatrixTest.cpp:260]:
```suggestion
TEST(MatrixTest, PatternClassification) {
```

File: mlir-sycl/unittests/Analysis/MemoryAccessAnalysis/MatrixTest.cpp [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+  builder.setInsertionPointToStart(funcOp.addEntryBlock());
+
+  Value zero = builder.create<arith::ConstantIntOp>(loc, 0, 32);
+  Value one = builder.create<arith::ConstantIntOp>(loc, 1, 32);
+  Value negativeOne = builder.create<arith::ConstantIntOp>(loc, -1, 32);
+  Value negativeTwo = builder.create<arith::ConstantIntOp>(loc, -2, 32);
+  Value two = builder.create<arith::ConstantIntOp>(loc, 2, 32);
+  auto bufferTy = MemRefType::get({}, builder.getI32Type());
+  Value memref = builder.create<memref::AllocaOp>(loc, bufferTy);
+  auto loadOp = builder.create<AffineLoadOp>(loc, memref);
```
Comment (whitneywhtsang) [inline mlir-sycl/unittests/Analysis/MemoryAccessAnalysis/MatrixTest.cpp:276]: What's the purpose of these operations?

**Review by whitneywhtsang — COMMENTED** [review]
comments on test case.

File: mlir-sycl/unittests/Analysis/MemoryAccessAnalysis/MatrixTest.cpp [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+  Value one = builder.create<arith::ConstantIntOp>(loc, 1, 32);
+  Value two = builder.create<arith::ConstantIntOp>(loc, 2, 32);
+
+  MemoryAccessMatrix matrix(3, 3);
+  matrix.fillRow(0, zero);
+  matrix.fillRow(1, one);
+  matrix.fillRow(2, two);
+
+  {
+    MemoryAccessMatrix subMatrix = matrix.getRows({0, 2});
```
Comment (whitneywhtsang) [inline mlir-sycl/unittests/Analysis/MemoryAccessAnalysis/MatrixTest.cpp:165]: can we also test that rows don't need to be increasing order, like `getRows({2, 0})`

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+  void setColumn(unsigned col, ArrayRef<Value> elems);
+
+  /// Fill \p col with the given value \p val.
+  void fillColumn(unsigned col, Value val);
+
+  /// Fill the matrix with the given value \p val.
+  void fill(Value val);
+
+  /// Construct a new matrix containing the specified \p rows.
+  MemoryAccessMatrix getRows(std::set<unsigned> rows) const;
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:187]: I would expect `getRows({2, 0})` to return the third row first, then the first row, using `std::set` would be wrong in that case, how about using `SmallSetVector`?

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+class AffineLoadOp;
+class AffineStoreOp;
+class DataFlowSolver;
+
+namespace sycl {
+
+/// Classify array access patterns.
+enum MemoryAccessPattern {
+  Unkown = 0,
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:32]:
```suggestion
  Unknown = 0,
```

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+  SmallVector<Value> getRow(unsigned row) const;
+
+  /// Set the specified \p row to \p elems.
+  void setRow(unsigned row, ArrayRef<Value> elems);
+
+  /// Fill \p row with the given value \p val.
+  void fillRow(unsigned row, Value val);
+
+  /// Add an extra row at the bottom of the matrix.
+  unsigned appendRow();
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:168]: What is the returned unsigned means?

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+  /// Construct a new matrix containing the sub-matrix specified by \p rows and
+  /// \p columns.
+  MemoryAccessMatrix getSubMatrix(std::set<unsigned> rows,
+                                  std::set<unsigned> columns) const;
+
+  //===----------------------------------------------------------------------===//
+  // Shape Queries
+  //===----------------------------------------------------------------------===//
+
+  /// Returns true if the matrix has equal number of rows and columns.
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:201]: nit: as the previous comments uses `Construct`, `Add`, `Get` without `s`, let's change all these `Returns` to `Return`.

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+  /// Returns true if the matrix is the filled with zero values.
+  bool isZero(DataFlowSolver &solver) const;
+
+  /// Returns true if the only non-zero entries are on the diagonal.
+  bool isDiagonal(DataFlowSolver &solver) const;
+
+  /// Returns true if the matrix is the unit matrix.
+  bool isIdentity(DataFlowSolver &solver) const;
+
+  /// Returns true if all non-zero entries are below the diagonal.
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:213]: How about the diagonal? The current implementation checks that they are non-zero entries too.

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+  bool hasStridedOverlappedAccessPattern(DataFlowSolver &solver) const;
+
+  /// Array accessed in strided fashion, decreasing offset order, starting at
+  /// different row offsets. The matrix shape is:
+  ///   |1  0|
+  ///   |1 -C|  where -C < -1
+  bool hasReverseStridedOverlappedAccessPattern(DataFlowSolver &solver) const;
+
+private:
+  /// Determine whether the value at \p row and \p column is a constant integer
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:270]:
```suggestion
  /// Return the value at \p row and \p column if it is a constant integer
```

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+
+  /// Set the specified \p row to the given \p offset value.
+  void setOffset(unsigned row, Value offset);
+
+  /// Fill the offset vector with the given value \p val.
+  void fill(Value val);
+
+  /// Add an extra element at the bottom of the offset vector and set it to the
+  /// given \p offset value.
+  unsigned append(Value offset);
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:339]: mention that the return value is the index?

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+  /// Returns true if the vector contains all zeros and the last element has
+  /// a strictly positive constant value.
+  bool isZeroWithLastElementStrictlyPositive(DataFlowSolver &solver) const;
+
+  /// Returns true if the vector contains all zeros and the last element has
+  /// value equal to the given constant \p k.
+  bool isZeroWithLastElementEqualTo(int k, DataFlowSolver &solver) const;
+
+private:
+  /// Determine whether the value at \p row is a constant integer value.
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:357]:
```suggestion
  /// /// Return the value at \p row if it is a constant integer value.
```

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+  const OpTy &accessOp;      /// The array load or store operation.
+  MemoryAccessMatrix matrix; /// The memory access matrix.
+  OffsetVector offsets;      /// The offset vector.
+};
+
+template <typename OpTy>
+inline raw_ostream &operator<<(raw_ostream &os,
+                               const MemoryAccess<OpTy> &access) {
+  os << "--- MemoryAccess ---\n\n";
+  os << "AccessMatrix:\n" << access.getAccessMatrix() << "\n";
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:412]: why not print the operation too?

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+using namespace mlir;
+using namespace mlir::sycl;
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Determine whether a value is an known integer value.
+static Optional<APInt> getConstIntegerValue(Value val, DataFlowSolver &solver) {
+  if (!isa<IntegerType>(val.getType()))
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp:27]: does this allow index type? how about 
```suggestion
  if (!val.getType().isIntOrIndex())
```

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp [hunk 16]
```diff
@@ -0,0 +1,652 @@
...
+
+    if (matrix.hasLinearOverlappedAccessPattern(solver))
+      return MemoryAccessPattern::LinearOverlapped;
+
+    if (matrix.hasStridedAccessPattern(solver))
+      return MemoryAccessPattern::Strided;
+
+    if (matrix.hasStridedOverlappedAccessPattern(solver))
+      return MemoryAccessPattern::StridedOverlapped;
+  } else {
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp:606]:
```suggestion
    return MemoryAccessPattern::Unkown;
  }
```

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+///   |      | * | | + |  |
+///   | 0  c3|   |j|   |c4|
+///
+template <typename OpTy> class MemoryAccess {
+  friend raw_ostream &operator<<(raw_ostream &, const MemoryAccess &);
+
+public:
+  MemoryAccess() = delete;
+
+  MemoryAccess(const OpTy &accessOp, MemoryAccessMatrix &&matrix,
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:390]: Technically, MemoryAccess should only take the accessOp, then matrix and offsets should be calculated?

**Review by victor-eds — COMMENTED** [review]
First pass just on the `.h` file

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h (hunk 10 above)
Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h]: ok I get it now, fixed.

File: mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h [hunk 18]
```diff
@@ -0,0 +1,434 @@
...
+               OffsetVector &&offsets);
+
+  const OpTy &getAccessOp() const { return accessOp; }
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h:405]:
```suggestion
  OpTy getAccessOp() const { return accessOp; }
```

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp [hunk 19]
```diff
@@ -0,0 +1,652 @@
...
+
+using namespace mlir;
+using namespace mlir::sycl;
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Determine whether a value is an known integer value.
+static Optional<APInt> getConstIntegerValue(Value val, DataFlowSolver &solver) {
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp:26]:
```suggestion
static std::optional<APInt> getConstIntegerValue(Value val, DataFlowSolver &solver) {
```

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp [hunk 20]
```diff
@@ -0,0 +1,652 @@
...
+/// Determine whether a value is equal to a given integer constant.
+static bool isEqualTo(Value val, int constant, DataFlowSolver &solver) {
+  Optional<APInt> optConstVal = getConstIntegerValue(val, solver);
+  if (!optConstVal.has_value())
+    return false;
+
+  APInt constVal = *optConstVal;
+  APInt c(constVal.getBitWidth(), constant, true /* signed */);
+  return (constVal == c);
+}
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp:52]:
```suggestion
static bool isEqualTo(Value val, uint64_t constant, DataFlowSolver &solver) {
  std::optional<APInt> optConstVal = getConstIntegerValue(val, solver);
  return optConstVal && (*optConstVal == constant);
}
```

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp [hunk 21]
```diff
@@ -0,0 +1,652 @@
...
+static bool isOne(Value val, DataFlowSolver &solver) {
+  return isEqualTo(val, 1, solver);
+}
+
+static bool isNegativeOne(Value val, DataFlowSolver &solver) {
+  return isEqualTo(val, -1, solver);
+}
+
+static bool isStrictlyPositive(Value val, DataFlowSolver &solver) {
+  Optional<APInt> constValue = getConstIntegerValue(val, solver);
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp:67]:
```suggestion
  std::optional<APInt> constValue = getConstIntegerValue(val, solver);
```
Use `std::optional` everywhere

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp [hunk 22]
```diff
@@ -0,0 +1,652 @@
...
+      Value val = at(row, col);
+      bool isOnDiagonal = (row == col);
+
+      // All values on the diagonal must be non-zero.
+      if (isOnDiagonal && ::isZero(val, solver))
+        return false;
+      // All other values must be zero.
+      if (!isOnDiagonal && !::isZero(val, solver))
+        return false;
+    }
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp:234]:
```suggestion
  for (size_t row = 0; row < nRows; ++row) {
    for (size_t col = 0; col < nColumns; ++col) {
      Value val = at(row, col);
      bool isOnDiagonal = (row == col);

      // All values on the diagonal must be non-zero.
      if (isOnDiagonal && ::isZero(val, solver))
        return false;
      // All other values must be zero.
      if (!isOnDiagonal && !::isZero(val, solver))
        return false;
    }
  }
```
This helps readability

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp [hunk 23]
```diff
@@ -0,0 +1,652 @@
...
+template <typename OpTy>
+MemoryAccess<OpTy>::MemoryAccess(const OpTy &accessOp,
+                                 MemoryAccessMatrix &&matrix,
+                                 OffsetVector &&offsets)
+    : accessOp(accessOp), matrix(std::move(matrix)),
+      offsets(std::move(offsets)) {
+  static_assert(std::is_same<AffineLoadOp, OpTy>::value ||
+                    std::is_same<AffineStoreOp, OpTy>::value,
+                "Expecting an affine load or store operation");
+}
```
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp]: I used `llvm::is_one_of` in the static_assert as I run into compilation issues changing the template class as you suggested.

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp (hunk 20 above)
Comment (victor-eds) [reply mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp]: Oh, that's interesting... LGTM in that case. Thanks!

Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp]: It doesn't work when int is negative e.g., -1.

File: mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp]: ok changed

**Review by victor-eds — APPROVED** [review]
